subprojects/core/src/main/java/org/gradle/api/internal/tasks/properties/bean/AbstractNestedRuntimeBeanNode.java

```diff
@@ -88,7 +88,13 @@ public Object create() {
                         }
                     }
                 });
-                return value instanceof Provider ? ((Provider<?>) value).getOrNull() : value;
+                // Replace absent Provider with null.
+                // This is required for allowing optional provider properties - all code which unpacks providers calls Provider.get() and would fail if an optional provider is passed.
+                // Returning null from a Callable is ignored, and PropertyValue is a callable.
+                if (value instanceof Provider && !((Provider<?>) value).isPresent()) {
```

Comment:
IIUC, I think we need a test that shows what happens when an optional output is consumed as a non-optional input.  Do we get a sensible error?

Skeleton of that:
```
class Producer extends DefaultTask {
    @Optional @OutputFile
    RegularFileProperty outputFile = newOutputFile() // never set
}
class Consumer extends DefaultTask {
    @Nested
    NestedInputFile nestedInput = new NestedInputFile(newInputFile())
}
class NestedInputFile {
   @InputFile
   final RegularFileProperty inputFile 
}
consumerTask.nestedInput.inputFile = producer.outputFile
```

This build should fail because `outputFile` isn't set, but I think the error will be something about `inputFile`.

How does this compare to the non-`@Nested` version?

Reply:
Why use `newInputFile()`? I think we should not use `newInputFile()` for properties in nested beans, only `layout.fileProperty()`. WDYT?

Reply:
I added the test. It fails with
```
No value has been specified for property 'bean.inputFile'.
```

I think that is perfectly fine. There is no problem with the configuration of the consumer task, right? It would be nice to see where the "value" of the input property comes from, like we had with the new model. But I don't think we have anything like this yet.